[NEW] frontend: Add toggle wrap for logViewer and fix log viewer styles

vyncent-t opened the pull request:
Fork PR from original: https://github.com/kubernetes-sigs/headlamp/pull/3027

requested feature from https://github.com/headlamp-k8s/headlamp/issues/2657

This PR adds a toggle button for the Logs Viewer window that allows the text to be formatted for word wrap. There are also style formatting issues that I have noticed and made an issue for here https://github.com/headlamp-k8s/headlamp/issues/3052  this PR also adds changes that style error while also. 



## Changes
![image](https://github.com/user-attachments/assets/d77ce163-5f73-4d02-8d46-847d738e6da0)
- **Introduces new Wrap option for log viewer component to allow toggle between default text formatting and word wrap**

![image](https://github.com/user-attachments/assets/2cf0202b-1199-432a-bb81-11a4b47db4fc)

- **Introduces new format and tooltip context for log viewer buttons** 
  - At current, adding additional switches or buttons to the log viewer components for both pod and non pod log viewer introduces styling errors (this PR makes it so that we can continue to add more features by reworking the button layout)

- **FIxes rendering for small view**
  - At current when viewing log viewer on smaller windows the buttons are both misaligned and not styled correctly https://github.com/headlamp-k8s/headlamp/issues/3052


### How to test

- Open headlamp into a cluster 
- Navigate to pods
- Click on any pod
- Open the "View logs" button
- Click the toggle "Wrap" button
- Log text format should change

joaquimrocha commented on frontend/src/components/common/LogViewer.tsx:
```diff
@@ -26,6 +26,38 @@ export interface LogViewerProps extends DialogProps {
+  isWrappedTextEnabled?: boolean;
+}
+
+// // this function manually formats the logs to a single line.
```
Why two // ?

vyncent-t replied:
I've moved on from using landmarks like "notez" for project wide search to find notes to self, forgot to clean this up, need to make this a /* comment with a note:

joaquimrocha commented on frontend/src/components/common/LogViewer.tsx:
```diff
@@ -26,6 +26,38 @@ export interface LogViewerProps extends DialogProps {
+}
+
+// // this function manually formats the logs to a single line.
+// // currently we must manually format the texts as addonFit does not work with multiline texts.
```
Can you elaborate on this? I don't understand exactly why this function is needed. Are you sure there's no way to delegate this to the xterm component?

vyncent-t replied:
I have tried using options for single line printing but it had a few issues.

the format to single line is needed since the log "lines" are not actually single line text strings but rather a bundle of log lines, which I found the terminal has trouble formatting it correctly. It also dynamically sets the width depending on the longest singular line, which that value I use to fix the width sizing.

![image](https://github.com/user-attachments/assets/39f5af03-b2e2-4e98-9ac0-4654858dbfb9)

(while the split new line \n is read properly by the console log, in the xterm it was not)
logs w/no timestamp

![image](https://github.com/user-attachments/assets/d2a5bab0-6ce1-48a8-ac09-cc3a2c49f5ad)

vyncent-t replied:
side note: I was responding in the context of you meaning "delegate" to the xterm comp as in using the built in options to handle the single line format printing (changing width, using addonFit, etc.) but if you mean in terms of moving this function from a stand alone into the main log viewer component I moved it there as well.

joaquimrocha commented on e2e-tests/tests/headlampPage.ts:
Is this commit related to the logs?

vyncent-t replied:
Not directly, it was needed for the e2e tests to pass initially but I am working on a fix for this in a e2e test related PR
https://github.com/kubernetes-sigs/headlamp/pull/3144 

this branch (along with a few others) have been seeing some e2e test failures mostly from the new multi cluster changes.

I was work shopping that potential fix and testing to see if the test failures were due to changes made for the logs (they were not, I introduced a fix into that PR, once that is merged in I can rebase and drop it from this branch)

vyncent-t replied:
actually, I will drop it now, just need to see if the e2e pass again

joaquimrocha reviewed (changes requested):
Besides my comments, I noticed that when we wrap the text, the vertical scroll bar is now pushed beyond the limits of the dialog, meaning we lose the ability to understand where we are vertically.

vyncent-t commented:
e2e tests are passing, none of the log changes cause it to fail. going to drop e2e test changes here and rebase when https://github.com/kubernetes-sigs/headlamp/pull/3144 is merged

sniok commented on frontend/src/components/common/LogViewer.test.ts:
```diff
@@ -0,0 +1,24 @@
+
+  it('returns 0 for empty input', () => {
+    expect(findLongestLine([])).toBe(0);
+  });
```
please add a test case like `["abcde", "egf\nabc"]`, I think the current implementation doesn't handle a case like that

vyncent-t replied:
I think 'splits on real LF (\\n) for a line' covers it since the new version of the findLongestLine uses `.split(/\n|\n|\r/)` so should split as needed but I added an another one with ["abcde", "egf\nabc"]

sniok commented on frontend/src/components/common/LogViewer.test.ts:
```diff
@@ -0,0 +1,28 @@
+  });
+
+  it('splits on real LF (\\n) for a line - longer', () => {
+    expect(findLongestLine(['abcde', 'egf\nabc'])).toBe(5 + 1);
```
the longest line in this case is `abcdeefg` and is 7 characters long

vyncent-t replied:
there isn't any joining after the function finds a new line or return syntax, it treats every "main string" (bundle of logs in live) as a collection that needs to be split up into parts and substrings and we count off those smaller parts

the inputs of ['abcde', 'egf\nabc'] are broken into ['abcde', 'egf' abc']

we then add a +1 on every line for the `longest` to make up for splitting the \n (it also makes sure the width is set to exactly the number of chars +1 to ensure a fit every time)

so the function should read 
`abcde` as 6 long
`efg` as 4 long
`abc` as 4 long

sniok replied:
yes I understand that currently the function works this way, but what I'm saying is that the function needs to be adjusted so it passes this test since it something we might encounter and it should properly count the line length

sniok reviewed (changes requested):
Wrap toggle is inverted, it wraps lines when the toggle is off

![image](https://github.com/user-attachments/assets/7bb04256-4390-4a77-a343-cf38a47dad57)

There's a lot of extra space at the bottom 
![image](https://github.com/user-attachments/assets/ea0ffc64-e75d-4f86-bb6f-612d4d305820)

All the toggles can should be placed here on wide screens
![image](https://github.com/user-attachments/assets/8bec2540-1d34-4172-a12b-50354445cb59)

vyncent-t commented:
> Wrap toggle is inverted, it wraps lines when the toggle is off

> All the toggles can should be placed here on wide screens ![image](https://private-user-images.githubusercontent.com/4517681/447076352-8bec2540-1d34-4172-a12b-50354445cb59.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDgwMTYxNTUsIm5iZiI6MTc0ODAxNTg1NSwicGF0aCI6Ii80NTE3NjgxLzQ0NzA3NjM1Mi04YmVjMjU0MC0xZDM0LTQxNzItYTEyYi01MDM1NDQ0NWNiNTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDUyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA1MjNUMTU1NzM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWI2NTIwZTU5Y2Q2YjI1MjI2NWM2MmVmYzEyODRiNjhkOGNlZjg2MWVhNWJmNzVjZjliNTRmZGIwN2MxNWZkOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.15WtyoS784Uz4Gb_wE5N5U8b6Wmbv6IV4gn5U_7Khc0)

- [x] fixed toggle inverse, since the logs currently load in in wrapped format I have fixed it to start as "Wrap lines" enabled
- [ ] WIP: fixing the extra space
- [x] Need feedback: All toggles and drop downs on one line in wide view
  - I have addressed a few issues in the current format where all toggles and drop downs are all on the same line in this issue https://github.com/headlamp-k8s/headlamp/issues/3052 . The changes I have currently not only fix the rendering issues, but also make things more future proof for adding more inputs later. I noticed that pods and deployments log viewer are different so by separating the drop downs above the toggles we can keep the look uniform.
  
**Pod view logs**
![image](https://github.com/user-attachments/assets/1e984cc4-2384-4d24-8520-e6efd296be4f)

**Other view logs**
![image](https://github.com/user-attachments/assets/8a2506b4-612c-49bc-833c-80c7ea7b5dc5)

Notice that by putting them all on one line here, we start to collide into the side icon buttons, preventing us from adding another button on the same line in the future.

vyncent-t commented:
new issues
- drawer mode: background does not flex fit when clicking one log, turning off wrap mode, then clicking another pod log to view it

vyncent-t commented:
last push:

- [x] moved to mix of grid and box for new display

vyncent-t commented:
May need to settle with the bottom spacing imperfections when not using wrap texts (toggling wrap text off) for some log windows.

This is only an issue for logs that do not contain enough lines to fill out the rest of the spacing when reformatting to unwrap.

I have also tried experimenting with creating a custom scroll bar just for this view and that introduces too many issues for what it is trying to accomplish (having to sync it to the 'real' scroll bar, having format issues with other switches due to needing nested boxes, etc.) and will be abandoning that idea. 

Although I have been able to succeed in making the unwrap view work along with the Prettify option where it preserves the JSON readable format but with unwrapped lines.

sniok commented:
xterm seems to be way limited, I think we should consider some alternative instead, that's more geared towards displaying logs instead of terminal emulator
